### ISSUE TITLE
Wrapping CLASS or CAMB

### ISSUE THREAD

**Issue by EiffL** [DifferentiableUniverseInitiative/jax_cosmo#23]
After a little bit of digging, it would technically be possible to fully wrap either code in `jax-cosmo`,  using finite difference to get an approximation of gradients. Just for future reference, or if someone wants to do it, here is the procedure:

  - [ ] The easy part, adding a new JAX primitive,  essentially, just telling JAX how to manipulate your function, see here: https://jax.readthedocs.io/en/latest/notebooks/How_JAX_primitives_work.html#Defining-new-JAX-primitives .  You can define a custom JVP based on finite differences, so that you can get approximate gradients within the rest of the computation. 

  - [ ] The not so easy part: Adding an XLA custom op to perform the computation.  So the problem is that doing the above will work, but you can't JIT  your program. For that, you would need to define a custom XLA op in C++ that would call CLASS/CAMB. It's technically possible, but painful >.< .... Here is the XLA reference for custom CPU ops: https://www.tensorflow.org/xla/custom_call

**Comment by EiffL** [comment]
I'm labeling this as won't fix, because although in principle possible, may turn out to be a major pain in the long run. So, I"m gonna wait until someone write their own Boltzmann code in JAX, and failing that, we'll probably use an emulator strategy!

**Comment by EiffL** [comment]
Oh wow, forget what I said:
https://github.com/google/jax/blob/master/jax/experimental/host_callback.py

With this API it looks like it will be possible to define host callbacks from compiled code!
